### ISSUE TITLE
[CLOSED] shallow-deep corres proof takes 3 hours

### ISSUE THREAD

**Issue by zilinc** [NICTA/cogent#190]
<a href="https://github.csiro.au/che283"><img src="https://github.csiro.au/avatars/u/286?" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [che283](https://github.csiro.au/che283)**
_Mon Jun 15 16:03:07 2015_
_Originally opened as https://github.csiro.au/ts-filesystems/Cogent/issues/85_

----

I don't have any further info right now as everything is on lemma...

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/kle060"><img src="https://github.csiro.au/avatars/u/248?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kle060](https://github.csiro.au/kle060)**
_Mon Jun 15 16:43:56 2015_

----

Thanks ;-)

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/kle060"><img src="https://github.csiro.au/avatars/u/248?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kle060](https://github.csiro.au/kle060)**
_Wed Jun 24 13:19:37 2015_

----

Narrowing down where the time is going:

These 2 are examples of taking way more time than they should in a sequential run:
scorres_get_page_and_find_space (67589.486s)
scorres_fsop_dir_rename_remove_old_dir (5704.598s)

This one was still running when I killed the sequential run after 24h:
scorres_ext2_new_block2

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/lim060"><img src="https://github.csiro.au/avatars/u/271?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lim060](https://github.csiro.au/lim060)**
_Mon Jul 27 10:53:52 2015_

----

Fun times for BilbyFS (`--formalisation=knf`) as of 81dbfea0d (with appropriate changes to the `let` tactic):

```
scorres_ostore_write_super: 53447.480s elapsed
scorres_fsop_do_rename_del: 51324.722s elapsed
scorres_fsop_mkdir: 43162.184s elapsed
scorres_fsop_create: 39573.891s elapsed
scorres_fsop_do_rename: 39421.007s elapsed
scorres_mount_recover: 37617.782s elapsed
scorres_fsop_symlink: 35227.360s elapsed
scorres_fsop_do_move_del: 31771.605s elapsed
scorres_ostore_sync: 27151.899s elapsed
scorres_fsop_move: 25239.418s elapsed
scorres_fsop_do_move: 23948.325s elapsed
scorres_ostore_read_sum: 14787.869s elapsed
scorres_fsop_rename: 14521.584s elapsed
scorres_fsop_link: 13890.421s elapsed
scorres_fsop_rmdir: 13665.556s elapsed
scorres_fsop_unlink: 13101.657s elapsed
scorres_fsop_init: 11623.234s elapsed
scorres_prepare_wbuf: 10737.204s elapsed
[...]
```

This is post-`XTHEN`, so there is some backtracking at the top level (on `head_tac`) or blow-up in the simplifier steps.

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/kle060"><img src="https://github.csiro.au/avatars/u/248?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kle060](https://github.csiro.au/kle060)**
_Tue Jul 28 14:41:03 2015_

----

It's probably the simp/clarsimp stage in solve_valRel, there shouldn't really be any backtracking going on any more. The log file should have output for each of the tactic stages (it would now be interleaved between different proofs, but you should still be able to see which ones are the major time sinks).

I haven't thought much yet about a good proof procedure for this step, I was mainly hoping that a clarsimp hammer would just do it. Looks like that is too slow.

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/lim060"><img src="https://github.csiro.au/avatars/u/271?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lim060](https://github.csiro.au/lim060)**
_Tue Aug  4 18:28:49 2015_

----

Confirmed; preliminary profiling results

```
3315.795 [atom_tac/con]
3300.872 [solve_valRel_tac/clarsimp 1]
3272.689 [atom_tac/con]
3258.494 [solve_valRel_tac/clarsimp 1]
2162.431 [atom_tac/step]
2156.258 [solve_valRel_tac/clarsimp 1]
1568.829 [atom_tac/con]
1567.370 [solve_valRel_tac/clarsimp 1]
1395.436 [atom_tac/con]
1393.663 [solve_valRel_tac/clarsimp 1]
1379.079 [atom_tac/step]
1372.834 [solve_valRel_tac/clarsimp 1]
1369.765 [atom_tac/con]
1368.287 [solve_valRel_tac/clarsimp 1]
1275.970 [atom_tac/step]
1272.196 [solve_valRel_tac/clarsimp 1]
1191.013 [atom_tac/step]
1173.225 [atom_tac/step]
1134.639 [solve_valRel_tac/clarsimp 1]
1133.545 [atom_tac/con]
1132.267 [solve_valRel_tac/clarsimp 1]
1118.028 [solve_valRel_tac/clarsimp 1]
995.986 [atom_tac/struct]
986.104 [atom_tac/struct]
984.825 [atom_tac/struct]
983.497 [solve_valRel_tac/clarsimp 1]
975.438 [solve_valRel_tac/clarsimp 1]
974.147 [solve_valRel_tac/clarsimp 1]
967.822 [head_tac/case]
960.470 [solve_valRel_tac/clarsimp 1]
949.160 [atom_tac/con]
947.448 [solve_valRel_tac/clarsimp 1]
887.142 [atom_tac/con]
886.359 [solve_valRel_tac/clarsimp 1]
857.304 [atom_tac/struct]
851.626 [atom_tac/struct]
836.229 [solve_valRel_tac/clarsimp 1]
834.471 [atom_tac/struct]
833.232 [solve_valRel_tac/clarsimp 1]
831.780 [solve_valRel_tac/clarsimp 1]
[...]
```

Main observations:
1. The `solve_valRel_tac` running time is dominated by `clarsimp 1` (unfolds records) instead of `clarsimp 2` (unfolds sums).
2. The `atom_tac` running time is dominated by `Con`, `Struct` and `step` (which includes `Put` and `Promote`, among other things). But nearly all of the time is spent in `solve_valRel_tac`.

Either way, I'm writing a new `scorres_tac` that will avoid unfolding `valRel` altogether (outside of the preprocessing stage).

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/lim060"><img src="https://github.csiro.au/avatars/u/271?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lim060](https://github.csiro.au/lim060)**
_Thu Aug  6 17:12:30 2015_

----

Fixed by 4dc25e5ce.

```
Timing Ext2_SCorres_Normal (4 threads, 562.294s elapsed time, 1687.232s cpu time, 208.440s GC time, factor 3.00)
Timing BilbyFS_SCorres_Normal (4 threads, 997.624s elapsed time, 3236.000s cpu time, 227.724s GC time, factor 3.24)
```
